managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/MetaStoreImpl.java

```diff
@@ -322,32 +334,99 @@ private static MetaStoreException getException(Throwable t) {
+        }
+        MLDataFormats.ManagedLedgerInfoMetadata mlInfoMetadata = MLDataFormats.ManagedLedgerInfoMetadata
+                .newBuilder()
+                .setCompressionType(ledgerInfoCompressionType)
+                .setUncompressedSize(managedLedgerInfo.getSerializedSize())
+                .build();
+        return compressManagedInfo(managedLedgerInfo.toByteArray(), mlInfoMetadata.toByteArray(),
+                mlInfoMetadata.getSerializedSize(), ledgerInfoCompressionType);
+    }
+
+    public byte[] compressCursorInfo(ManagedCursorInfo managedCursorInfo) {
+        if (cursorInfoCompressionType.equals(CompressionType.NONE)) {
+            return managedCursorInfo.toByteArray();
+        }
+        MLDataFormats.ManagedCursorInfoMetadata metadata = MLDataFormats.ManagedCursorInfoMetadata
+                .newBuilder()
+                .setCompressionType(cursorInfoCompressionType)
+                .setUncompressedSize(managedCursorInfo.getSerializedSize())
+                .build();
+        return compressManagedInfo(managedCursorInfo.toByteArray(), metadata.toByteArray(),
+                metadata.getSerializedSize(), cursorInfoCompressionType);
+    }
+
+    public ManagedLedgerInfo parseManagedLedgerInfo(byte[] data) throws InvalidProtocolBufferException {
+        ByteBuf byteBuf = Unpooled.wrappedBuffer(data);
+
+        byte[] metadataBytes = extractCompressMetadataBytes(byteBuf);
+        if (metadataBytes != null) {
+        try {
+                MLDataFormats.ManagedLedgerInfoMetadata metadata =
+                        MLDataFormats.ManagedLedgerInfoMetadata.parseFrom(metadataBytes);
+                byte[] decodeBytes =
+                        parseCompressedInfo(byteBuf, metadata.getCompressionType(), metadata.getUncompressedSize());
+                return ManagedLedgerInfo.parseFrom(decodeBytes);
+            } catch (Exception e) {
+                log.error("Failed to parse managedLedgerInfo metadata, "
+                        + "fall back to parse managedLedgerInfo directly.", e);
+                return ManagedLedgerInfo.parseFrom(data);
+            } finally {
+                byteBuf.release();
+            }
+        } else {
+            return ManagedLedgerInfo.parseFrom(data);
+        }
+    }
+
+    public ManagedCursorInfo parseManagedCursorInfo(byte[] data) throws InvalidProtocolBufferException {
+        ByteBuf byteBuf = Unpooled.wrappedBuffer(data);
+
+        byte[] metadataBytes = extractCompressMetadataBytes(byteBuf);
+        if (metadataBytes != null) {
+            try {
+                MLDataFormats.ManagedCursorInfoMetadata metadata =
+                        MLDataFormats.ManagedCursorInfoMetadata.parseFrom(metadataBytes);
+                byte[] decodeBytes =
+                        parseCompressedInfo(byteBuf, metadata.getCompressionType(), metadata.getUncompressedSize());
+                return ManagedCursorInfo.parseFrom(decodeBytes);
```

Comment:
ManagedCursorInfo.parseFrom() already support parse `java.nio.ByteBuffer`

```java
    public static org.apache.bookkeeper.mledger.proto.MLDataFormats.ManagedLedgerInfo parseFrom(
        java.nio.ByteBuffer data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
```

We don't need to parse to byte array again?

Reply:
This ByteBuf is provided by netty, not java.

Reply:
@codelipenghui

Reply:
@nodece `ByteBuf.nioBuffer()` can help